### PR TITLE
feat(tauri) lazy loading for the no-server mode

### PR DESCRIPTION
**Pull request by lucasfernog**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] New Binding Issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes. Issue #___
- [x] No

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch and _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

File: cli/tauri.js/runner.js [hunk 1]
```diff
@@ -110,6 +121,34 @@ class Runner {
...
+        if (err) {
+          reject(err)
+        } else {
+          const dom = new JSDOM(html)
+          const document = dom.window.document
+          document.querySelectorAll('link').forEach(link => {
+            link.removeAttribute('rel')
+            link.removeAttribute('as')
+          })
+
```
Comment (nothingismagick) [inline cli/tauri.js/runner.js]: Brilliant.

File: cli/tauri.js/runner.js [hunk 2]
```diff
@@ -110,6 +121,34 @@ class Runner {
...
+          const tauriScript = document.createElement('script')
+          tauriScript.text = readFileSync(path.join(tauriDir, 'tauri.js'))
+          document.body.insertBefore(tauriScript, document.body.firstChild)
+
```
Comment (nothingismagick) [inline cli/tauri.js/runner.js]: And there's the secret sauce.

File: cli/tauri.js/templates/tauri.js [hunk 3]
```diff
@@ -366,28 +363,50 @@ window.tauri = {
...
-      var target = e.target;
+  loadAsset: function loadAsset(assetName) {
+    return this.promisified({
+      cmd: 'loadAsset',
+      asset: assetName
+    })
+  }
+};
+
+window.addEventListener('message', function (event) {
```
Comment (nothingismagick) [inline cli/tauri.js/templates/tauri.js]: We might want to be able to properly tree shake this out if they don't enable it.

File: examples/vue/quasar-app/tauri.conf.js [hunk 4]
```diff
@@ -10,7 +10,7 @@ module.exports = function () {
     ctx: {},
     tauri: {
       embeddedServer: {
-        active: true
+        active: false
```
Comment (nothingismagick) [inline examples/vue/quasar-app/tauri.conf.js]: 🐉 🔥 🥇

File: tauri/src/api.rs [hunk 5]
```diff
@@ -16,27 +18,20 @@ pub fn handler<T: 'static>(webview: &mut WebView<'_, T>, arg: &str) -> bool {
               _webview
                 .eval(&format!(
                   "window['{queue}'] = [];
-                  window['{fn}'] = function (payload, salt, ignoreQueue) {{
-                    window.tauri.promisified({{
-                      cmd: 'validateSalt',
-                      salt: salt
-                    }}).then(function () {{
```
Comment (nothingismagick) [inline tauri/src/api.rs]: Will this be coming back?

File: webpack/plugins/tauri-require.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
+module.exports.plugin = class TauriRequirePlugin {
+  apply (compiler) {
+    compiler.plugin('compilation', function (compilation) {
+      compilation.mainTemplate.hooks.requireEnsure.tap('JsonpMainTemplatePlugin load', (source, chunk, hash) => {
+        return `
+          // Tauri chunk loading
+
+          var installedChunkData = installedChunks[chunkId];
```
Comment (nothingismagick) [inline webpack/plugins/tauri-require.js]: are you linting? there's a lot of stuff here that I don't like
var, ;

File: cli/tauri.js/templates/tauri.js (hunk 3 above)
Comment (lucasfernog) [reply cli/tauri.js/templates/tauri.js]: oops, this should not be here anymore lol

File: webpack/plugins/tauri-require.js (hunk 6 above)
Comment (lucasfernog) [reply webpack/plugins/tauri-require.js]: I just adapted it from webpack's code

File: tauri/src/api.rs (hunk 5 above)
Comment (lucasfernog) [reply tauri/src/api.rs]: yeah, and we also have to think about locking every API call with salt

File: cli/tauri.js/runner.js (hunk 1 above)
Comment (lucasfernog) [reply cli/tauri.js/runner.js]: had to do this, because the webview won't load things with rel=preload

**Comment by nothingismagick** [comment]
A couple notes:
1. The quasar example with no-server works (even if the bundle is a little large)
2. I've got two warnings when I build:
```
warning: unused macro definition
  --> /[redacted]/tauri/tauri/src/macros.rs:22:1
   |
22 | / macro_rules! print_flush {
23 | |     ($literal:expr) => {
24 | |         print!($literal);
25 | |         ::std::io::Write::flush(&mut ::std::io::stdout())?;
...  |
30 | |     }
31 | | }
   | |_^
   |
   = note: `#[warn(unused_macros)]` on by default

warning: unused macro definition
  --> /[redacted]/tauri/tauri/src/macros.rs:34:1
   |
34 | / macro_rules! set_ssl_vars {
35 | |   () => {
36 | |     #[cfg(target_os = "linux")]
37 | |     {
...  |
45 | |   };
46 | | }
   | |_^
```

**Comment by lucasfernog** [comment]
The bundle is a little large because right now it's bundling the CSS twice. This is going to be a little tricky to fix.

**Comment by nklayman** [comment]
When `embeddedServer.active = false`, images fail to load with the vue cli plugin, but they work fine if it is `true`.

**Pull request merged**